caikit/runtime/http_server.py

```diff
@@ -0,0 +1,477 @@
+    StatusCode.RESOURCE_EXHAUSTED: 429,
+    StatusCode.CANCELLED: 499,
+    StatusCode.UNKNOWN: 500,
+    StatusCode.DATA_LOSS: 500,
+    StatusCode.UNIMPLEMENTED: 501,
+    StatusCode.UNAVAILABLE: 501,
+    StatusCode.DEADLINE_EXCEEDED: 504,
+}
+
+
+# These keys are used to define the logical sections of the request and response
+# data structures.
+REQUIRED_INPUTS_KEY = "inputs"
+OPTIONAL_INPUTS_KEY = "parameters"
+
+## RuntimeHTTPServer ###########################################################
+
+
+class RuntimeHTTPServer(RuntimeServerBase):
+    """An implementation of a FastAPI server that serves caikit runtimes"""
+
+    ###############
+    ## Interface ##
+    ###############
+
+    def __init__(self, tls_config_override: Optional[aconfig.Config] = None):
+        super().__init__(get_config().runtime.http.port, tls_config_override)
+
+        self.app = FastAPI()
+
+        # Set up the central predict servicer
+        inference_service = ServicePackageFactory().get_service_package(
+            ServicePackageFactory.ServiceType.INFERENCE,
+        )
+        self.global_predict_servicer = GlobalPredictServicer(inference_service)
+
+        # Set up the central train servicer
+        # TODO: uncomment later on
+        # train_service = ServicePackageFactory().get_service_package(
+        #     ServicePackageFactory.ServiceType.TRAINING,
+        # )
+        # self.global_train_servicer = GlobalTrainServicer(train_service)
+
+        self.package_name = inference_service.descriptor.full_name.rsplit(".", 1)[0]
+
+        # Bind all routes to the server
+        self._bind_routes(inference_service)
+        # self._bind_routes(train_service)
+
+        # Parse TLS configuration
+        tls_kwargs = {}
+        if (
+            self.tls_config
+            and self.tls_config.server.key
+            and self.tls_config.server.cert
+        ):
+            log.info("<RUN10001905I>", "Running with TLS")
+            tls_kwargs["ssl_keyfile"] = self.tls_config.server.key
+            tls_kwargs["ssl_certfile"] = self.tls_config.server.cert
+            if self.tls_config.client.cert:
+                log.info("<RUN10001809I>", "Running with mutual TLS")
+                tls_kwargs["ssl_ca_certs"] = self.tls_config.client.cert
+                tls_kwargs["ssl_cert_reqs"] = ssl.CERT_REQUIRED
+
+        # Start the server with a timeout_graceful_shutdown
+        # if not set in config, this is None and unvicorn accepts None or number of seconds
+        unvicorn_timeout_graceful_shutdown = (
+            get_config().runtime.http.server_shutdown_grace_period_seconds
+        )
+        config = uvicorn.Config(
+            self.app,
+            host="0.0.0.0",
+            port=self.port,
+            log_level=None,
+            log_config=None,
+            timeout_graceful_shutdown=unvicorn_timeout_graceful_shutdown,
+            **tls_kwargs,
+        )
+        self.server = uvicorn.Server(config=config)
```

Comment:
I think we need to pre-declare `self.uvicorn_server_thread` (and maybe add a leading `_` to the name) so that line 184 below doesn't raise an `AttributeError` if running synchronously. I think this case could happen if you inverted the pattern here and rather than running the server in a thread, you launched a "shutdown listener" thread that called `stop` based on some signal it got.

```suggestion
        self.server = uvicorn.Server(config=config)
        
        # Placeholder for thread when running without blocking
        self._uvicorn_server_thread = None
```